src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp

```diff
@@ -232,18 +233,41 @@ void ShenandoahConcurrentMark::mark_concurrent_roots() {
   workers->run_task(&task);
 }
 
+class ShenandoahFlushSATBHandshakeClosure : public HandshakeClosure {
+private:
+  SATBMarkQueueSet& _qset;
+public:
+  ShenandoahFlushSATBHandshakeClosure(SATBMarkQueueSet& qset) :
+    HandshakeClosure("Shenandoah Flush SATB Handshake"),
+    _qset(qset) {}
+
+  void do_thread(Thread* thread) {
+    _qset.flush_queue(ShenandoahThreadLocalData::satb_mark_queue(thread));
+  }
+};
+
 void ShenandoahConcurrentMark::concurrent_mark() {
   ShenandoahHeap* const heap = ShenandoahHeap::heap();
   WorkGang* workers = heap->workers();
   uint nworkers = workers->active_workers();
   task_queues()->reserve(nworkers);
 
-  {
+  ShenandoahSATBMarkQueueSet& qset = ShenandoahBarrierSet::satb_mark_queue_set();
+  ShenandoahFlushSATBHandshakeClosure flush_satb(qset);
+  int enqueued_count_before;
+  int enqueued_count_after;
+  int flushes = 0;
+  int max_flushes = ShenandoahMaxSATBBufferFlushes;
+  do {
     TaskTerminator terminator(nworkers, task_queues());
     ShenandoahConcurrentMarkingTask task(this, &terminator);
     workers->run_task(&task);
-  }
 
+    enqueued_count_before = qset.completed_buffers_num();
+    Handshake::execute(&flush_satb);
+    enqueued_count_after = qset.completed_buffers_num();
+    flushes++;
+  } while (enqueued_count_before != enqueued_count_after && flushes < max_flushes);
```

Comment:
So, how does this interact with cancellation? Shouldn't we check for `cancelled_gc()` here as well?

Reply:
I think this would be cleaner:

```
  ShenandoahFlushSATBHandshakeClosure flush_satb(qset);

  for (int flushes = 0; flushes < ShenandoahMaxSATBBufferFlushes; flushes++) {
    TaskTerminator terminator(nworkers, task_queues());
    ShenandoahConcurrentMarkingTask task(this, &terminator);
    workers->run_task(&task);

    if (cancelled_gc()) {
      // GC is cancelled, break out.
      break;
    }

    int before = qset.completed_buffers_num();
    Handshake::execute(&flush_satb);
    int after = qset.completed_buffers_num();

    if (before == after) {
       // No more retries needed, break out.
       break;
     }
  }
```